FIPS/fip-draft_foundational_fvm.md

```diff
@@ -0,0 +1,1010 @@
+straightforward and relatively risk-free porting of existing battle-tested
+Ethereum smart contracts to the Filecoin network.
+
+Every foreign runtime is in turn an actor type, identified by a `CodeCID`.
+Foreign runtimes are installed just like any other actor type. Foreign actors
+instances are constructed through the `InitActor`, supplying the source
+deployable as a constructor.
+
+In the case of the EVM foreign runtime, we are looking to standardise on either
+one of these interpreters:
+
+1. [SputnikVM], or
+2. [blueallow/revm] (a fork of the above)
+
+Every foreign runtime will need to specify bindings to map runtime-specific
+aspects (e.g. addressing, crypto primitives, gas accounting, etc.) to Filecoin
+counterparts. A preliminary [EVM <> FVM binding specification] exists.
+
+Future foreign runtime candidates include Agoric SES, Solana eBPF, and others.
+In all cases, compatibility should target the lowest level executable output in
+its source form, rather than dealing with high-level languages using
+alternative/custom toolchains. This reduces risk surface by increasing execution
+fidelity/parity, and makes it tooling reuse possible.
+
+#### Actor bytecode representation
+
+WASM bytecode payloads will be represented as an IPLD DAG, and will be stored in
+the state blockstore, owned by the node implementation.
+
+**The concrete DAG layout, multihash function, and CID codec are yet to be
+defined.**
+
+Code will be linked to from the state tree through the existing `code` CID field
+in the `ActorState` object:
+
+```rust
+pub struct ActorState {
+    /// Link to code for the actor.
+    pub code: Cid,
+    /// Link to the state of the actor.
+    pub state: Cid,
+    /// Current sequential nonce of the actor.
+    pub nonce: u64,
+    /// Filecoin tokens available to the actor.
+    pub balance: TokenAmount,
+}
+```
+
+#### Actor deployment
+
+The current `InitActor` (`f00`) will be extended with a (`InstallActor`) method
+taking two input parameters:
+
+1. The actor's WASM bytecode as an input parameter.
+2. The invocation container version.
+
+The state of `InitActor` will be augmented with an `ActorRegistry` HAMT of `CID
+=> ActorSpec`, where `ActorSpec` is:
+
+```
+/// A simple actor specification/metadata structure, to be extended in the future.
+struct ActorSpec {
+  /// The version of the invocation container.
+  version: int, 
+}
+```
+
+The logic of `InitActor#LoadActor` is as follows:
+
+1. Validate the WASM bytecode.
+    - Syntax validation and structural validation, [as per standard](https://webassembly.github.io/spec/core/valid/index.html).
+    - No floating point instructions.
+2. Multihash the bytecode and calculate the code CID.
+3. Check if the code CID is already registered in the `ActorRegistry`.
+4. If yes, charge gas for the above operations and return the existing CID.
+5. Insert an entry in the `ActorRegistry`, binding the bytecode CID with the
```

Comment:
Could maybe simplify this by appealing to idempotency, but I'm not sure if that'll help.

Reply:
IMO mentioning idempotency here may lead to confusion because the _full_ logic is not idempotent due to the lower gas on duplicate.